GHermeneus/include/GHermeneus/Machine.h

```diff
@@ -5,4 +5,185 @@
 #ifndef GHERMENEUS_MACHINE_H
 #define GHERMENEUS_MACHINE_H
 
+#include <algorithm>
+#include <execution>
+#include <string>
+#include <string_view>
+#include <vector>
+#include <iostream>
+#include <mutex>
+#include <optional>
+
+#include <range/v3/distance.hpp>
+#include <range/v3/view/split.hpp>
+#include <range/v3/view/transform.hpp>
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/take.hpp>
+#include <range/v3/view/drop.hpp>
+#include <range/v3/range/operations.hpp>
+
+#include "GCode.h"
+#include "Instruction.h"
+#include "Parameters.h"
+
+namespace GHermeneus
+{
+    /*!
+     * This is the workhorse, here the GCode is processed and converted to different output types,
+     * such as protobufs, CSV text streams, Numpy objects.
+     *
+     * Actual implementation of this abstract type is usually done for each type GCode dialect. A Dialect is a namespace
+     * such as the GHermeneus::Dialects::Marlin. Here the Machine is defined as:
+     *
+     *      using MarlinSSV = StateSpaceVector<double, 10>; // n = t, x, y, z, x_dot, y_dot, z_dot, e, e_dot, T
+     *      using MarlinMachine = Machine<MarlinSSV, double>;
+     *
+     * Normal usage of a DialectMachine would then be:
+     *
+     *      using namespace GHermeneus::Dialects::Marlin;
+     *      auto UM3 = MarlinMachine();
+     *      UM3 << gcode;
+     *      std::cout << UM3;
+     *
+     * Which will output a CSV formatted array of the instructions (not yet implemented fully)
+     *
+     * @brief The workhorse behind this parser it will convert GCode to Instructions and different outputs
+     * @tparam SSV_T The type of the State Space Vector
+     * @tparam T  The primitive type of the State Space eq. double, int etc
+     */
+    template<typename SSV_T, typename T>
+    class Machine
+    {
+    public:
+
+        Machine() = default;
+
+        Machine(Machine<SSV_T, T>&& machine)  noexcept = default;
+
+        virtual ~Machine() = default;
+
+        /*!
+         * @brief parse the GCode to the machine instruction vector
+         * @param GCode a string_view containing the GCode
+         */
+        void parse(const std::string_view& GCode)
+        {
+            gcode = GCode;
+            extractLines(GCode); // extract the individual GCode lines
+
+            // Extract the Commands from each line this is done in an parallel unsequenced execution strategy
+            cmdlines.reserve(lines.size());
+            std::for_each(std::execution::par_unseq, lines.begin(), lines.end(), [&](Line& line) {
+                std::lock_guard<std::mutex> guard(cmdlines_mutex);
```

Comment:
This'll lock the mutex in each iteration of the for-loop, not only completely negating the parallel execution, but also giving some overhead.

It would make more sense to place the lock-guard/mutex _within_ the if statement, so at least the `extractCmd(line)` bit can run in an actually parallel fashion.

However, even then ... This is almost identical to a 'the user is not allowed to do this' example on cppreference:
https://en.cppreference.com/w/cpp/algorithm/execution_policy_tag_t
(at the bottom of the page). Meaning you can make it parallel, but not unsequenced (I think).

Reply:
I knew I asked the right person to review it 😉 thanks this is valuable and I will change it.

Reply:
As discussed over slack, I will try zipping the lines and the cmdlines with ranges and dor a for_each par_unseq over that zipped range. I believe that even if we loose some overhead in the preparation for this par_unseq operation we will gain significant speed gains in the end. GCode files might contain millions of lines which don't need to be processed in sequence at this point in the program.

```C++
std::vector<Instruction<SSV_T, T>> cmdlines { lines.size(), Instruction<SSV_T, T>() };
auto zipped_lines = ranges::views::zip(lines, cmdlines) | ranges::to_vector;
std::for_each(std::execution::par_unseq, zipped_lines.begin(), zipped_lines.end(), [&](auto line) {
    if (auto extractedCmd = extractCmd(line.first))
   {
        line.second = extractedCmd.value();
   }});
```

Reply:
Nope, the proposed code doesn't seem to work. But I will prevail, I won't give up!